File: C++/maximum-score-from-removing-substrings.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    int maximumGain(string s, int x, int y) {
        string a = "ab", b = "ba";
        if (x < y) {
            swap(a, b);
            swap(x, y);
        }
        return remove(s, a, x) + remove(s, b, y);
    }

private:
    int remove(string &s, string a, int x) {
        int i = 0, result = 0;
        for (int j = 0; j < size(s); ++j) {
            s[i++] = s[j];
            if (i >= 2 && s.substr(i - 2, 2) == a) {
                i -= 2;
                result += x;
            }
        }
        s.resize(i);
        return result;
    }
};

File: Python/maximum-score-from-removing-substrings.py
# Time:  O(n)
# Space: O(n)

class Solution(object):
    def maximumGain(self, s, x, y):
        """
        :type s: str
        :type x: int
        :type y: int
        :rtype: int
        """
        def score(s, a, x):
            i = result = 0
            for j in xrange(len(s)):
                s[i] = s[j]
                i += 1
                if i >= 2 and s[i-2:i] == a:
                    i -= 2
                    result += x
            s[:] = s[:i]
            return result
                
        s, a, b = list(s), list("ab"), list("ba")
        if x < y:
            x, y = y, x
            a, b = b, a
        return score(s, a, x) + score(s, b, y)
